stock_picking_location_check/tests/test_stock_picking_location_check.py

```diff
@@ -0,0 +1,106 @@
+
+
+class TestStockPickingLocationCheck(TransactionCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.product = cls.env["product.product"].create({"name": "Test Product"})
+        cls.supplier_location = cls.env.ref("stock.stock_location_suppliers")
+        cls.customer_location = cls.env.ref("stock.stock_location_customers")
+        cls.shelf1 = cls.env.ref("stock.stock_location_components")
+        cls.other_location = cls.env["stock.location"].create(
+            {"name": "Other Location"}
+        )
+        # Create incoming and outgoing pickings
+        cls.picking_in = cls.env["stock.picking"].create(
+            {
+                "location_id": cls.supplier_location.id,
+                "location_dest_id": cls.shelf1.id,
+                "picking_type_id": cls.env.ref("stock.picking_type_in").id,
+                "allow_location_inconsistency": False,
+            }
+        )
+        cls.picking_out = cls.env["stock.picking"].create(
+            {
+                "location_id": cls.shelf1.id,
+                "location_dest_id": cls.customer_location.id,
+                "picking_type_id": cls.env.ref("stock.picking_type_out").id,
+                "allow_location_inconsistency": False,
+            }
+        )
+
+    def create_move_and_move_line(
+        self, picking, product, location_id, location_dest_id, qty
+    ):
+        move = self.env["stock.move"].create(
+            {
+                "name": product.name,
+                "product_id": product.id,
+                "product_uom": product.uom_id.id,
+                "location_id": location_id,
+                "location_dest_id": location_dest_id,
+                "picking_id": picking.id,
+                "quantity_done": qty,
+                "state": "assigned",
+            }
+        )
+        self.env["stock.move.line"].create(
+            {
+                "move_id": move.id,
+                "product_id": product.id,
+                "product_uom_id": product.uom_id.id,
+                "location_id": location_id,
+                "location_dest_id": location_dest_id,
+                "picking_id": picking.id,
+                "qty_done": qty,
+            }
+        )
+
+    def test_incoming_location_discrepancy_allowed(self):
+        self.picking_in.allow_location_inconsistency = True
+        self.create_move_and_move_line(
+            self.picking_in,
+            self.product,
+            self.supplier_location.id,
+            self.other_location.id,
+            10,
+        )
+        self.picking_in.button_validate()
+
+    def test_incoming_location_discrepancy_not_allowed(self):
+        self.create_move_and_move_line(
+            self.picking_in,
+            self.product,
+            self.supplier_location.id,
+            self.other_location.id,
+            10,
+        )
+        with self.assertRaises(UserError):
+            self.picking_in.button_validate()
```

Comment:
We should combine these two test.

```suggestion
    def test_incoming_location_discrepancy_check(self):
        self.create_move_and_move_line(
            self.picking_in,
            self.product,
            self.supplier_location.id,
            self.other_location.id,
            10,
        )
        with self.assertRaises(UserError):
            self.picking_in.button_validate()
        self.picking_in.allow_location_inconsistency = True
        self.picking_in.button_validate()
```

There is also not much point separating tests for incoming and outgoing, IMO. Shall we consolidate them as well?